File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/data/repository/ItemSaleTaxRepositoryImpl.java
package com.bits.bee.bpmc.data.repository;

import java.lang.System;

/**
 * Created by aldi on 20/05/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u00008\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u000e\n\u0000\n\u0002\u0018\u0002\n\u0002\u0010 \n\u0000\u0018\u00002\u00020\u0001B\u001f\b\u0007\u0012\u0006\u0010\u0002\u001a\u00020\u0003\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u00a2\u0006\u0002\u0010\bJ\u0018\u0010\t\u001a\n\u0012\u0006\u0012\u0004\u0018\u00010\u000b0\n2\u0006\u0010\f\u001a\u00020\rH\u0016J\u001a\u0010\u000e\u001a\u0014\u0012\u0010\u0012\u000e\u0012\n\u0012\b\u0012\u0004\u0012\u00020\u000b0\u00100\u000f0\nH\u0016R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0002\u001a\u00020\u0003X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006\u0011"}, d2 = {"Lcom/bits/bee/bpmc/data/repository/ItemSaleTaxRepositoryImpl;", "Lcom/bits/bee/bpmc/domain/repository/ItemSaleTaxRepository;", "defaultDispatcher", "Lkotlinx/coroutines/CoroutineDispatcher;", "itemSaleTaxDao", "Lcom/bits/bee/bpmc/data/data_source/local/dao/ItemSaleTaxDao;", "apiUtils", "Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;", "(Lkotlinx/coroutines/CoroutineDispatcher;Lcom/bits/bee/bpmc/data/data_source/local/dao/ItemSaleTaxDao;Lcom/bits/bee/bpmc/data/data_source/remote/ApiUtils;)V", "getItemTaxByCodeItem", "Lkotlinx/coroutines/flow/Flow;", "Lcom/bits/bee/bpmc/domain/model/ItemSaleTax;", "codeItem", "", "getLatestItemSaleTax", "Lcom/bits/bee/bpmc/utils/Resource;", "", "app_debug"})
public final class ItemSaleTaxRepositoryImpl implements com.bits.bee.bpmc.domain.repository.ItemSaleTaxRepository {
    private final kotlinx.coroutines.CoroutineDispatcher defaultDispatcher = null;
    private final com.bits.bee.bpmc.data.data_source.local.dao.ItemSaleTaxDao itemSaleTaxDao = null;
    private final com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils = null;
    
    @javax.inject.Inject()
    public ItemSaleTaxRepositoryImpl(@org.jetbrains.annotations.NotNull()
    kotlinx.coroutines.CoroutineDispatcher defaultDispatcher, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.local.dao.ItemSaleTaxDao itemSaleTaxDao, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.data.data_source.remote.ApiUtils apiUtils) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.utils.Resource<java.util.List<com.bits.bee.bpmc.domain.model.ItemSaleTax>>> getLatestItemSaleTax() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlinx.coroutines.flow.Flow<com.bits.bee.bpmc.domain.model.ItemSaleTax> getItemTaxByCodeItem(@org.jetbrains.annotations.NotNull()
    java.lang.String codeItem) {
        return null;
    }
}

File: app/src/main/java/com/bits/bee/bpmc/data/repository/ItemSaleTaxRepositoryImpl.kt
package com.bits.bee.bpmc.data.repository

import com.bits.bee.bpmc.data.data_source.local.dao.ItemSaleTaxDao
import com.bits.bee.bpmc.data.data_source.local.model.ItemSaleTaxEntity
import com.bits.bee.bpmc.data.data_source.remote.ApiUtils
import com.bits.bee.bpmc.data.data_source.remote.response.ItemBranchResponse
import com.bits.bee.bpmc.data.data_source.remote.response.ItemTaxResponse
import com.bits.bee.bpmc.domain.mapper.ItemBranchDataMapper
import com.bits.bee.bpmc.domain.mapper.ItemSaleTaxDataMapper
import com.bits.bee.bpmc.domain.model.ItemBranch
import com.bits.bee.bpmc.domain.model.ItemSaleTax
import com.bits.bee.bpmc.domain.repository.ItemSaleTaxRepository
import com.bits.bee.bpmc.utils.ApiResponse
import com.bits.bee.bpmc.utils.NetworkDatabaseBoundResource
import com.bits.bee.bpmc.utils.Resource
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import javax.inject.Inject

/**
 * Created by aldi on 20/05/22.
 */
class ItemSaleTaxRepositoryImpl @Inject constructor(
    private val defaultDispatcher: CoroutineDispatcher,
    private val itemSaleTaxDao: ItemSaleTaxDao,
    private val apiUtils: ApiUtils
) : ItemSaleTaxRepository {

    override fun getLatestItemSaleTax(): Flow<Resource<List<ItemSaleTax>>> {
        return object : NetworkDatabaseBoundResource<List<ItemSaleTax>, ItemTaxResponse>() {
            override suspend fun loadFormDB(): List<ItemSaleTax>? {
                return itemSaleTaxDao.getItemSaleTax().map { ItemSaleTaxDataMapper.fromDbToDomain(it) }
            }

            override fun shouldFetch(data: List<ItemSaleTax>?): Boolean {
                return true
            }

            override suspend fun createCall(): Flow<ApiResponse<ItemTaxResponse>> {
                return apiUtils.getItemApiService().getItemTax()
            }

            override suspend fun saveCallResult(data: ItemTaxResponse) {
                itemSaleTaxDao.insertBulk(data.data.map { ItemSaleTaxDataMapper.fromNetworkToDb(it) })
            }

        }.getAsFlow()
    }

    override fun getItemTaxByCodeItem(codeItem: String): Flow<ItemSaleTax?> = flow {
        emit(itemSaleTaxDao.getItemTaxByCodeItem(codeItem)?.let { ItemSaleTaxDataMapper.fromDbToDomain(it) })
    }.flowOn(defaultDispatcher)


}